src/inmanta/execute/dataflow/root_cause.py

```diff
@@ -72,31 +72,51 @@ def root_causes(self) -> Set[AttributeNode]:
             Rules 2 to 4 are implemented as propagation steps by
             _assignment_step, _child_attribute_step and _parent_instance_step respectively.
         """
-        causes: Set[AttributeNode] = set(())
-        for node in self.nodes:
-            others: FrozenSet[AttributeNode] = self.nodes.difference({node})
-            seen: Set[AssignableNode] = set(())
-            to_check: List[AssignableNode] = [node]
-
-            def process_step(step_result: FrozenSet[AssignableNode]) -> None:
-                new: Set[AssignableNode] = set(step_result).difference(seen)
-                to_check.extend(new)
-                seen.update(new)
-
-            is_root_cause: bool = True
-            while to_check:
-                n: AssignableNode = to_check.pop()
-                if n in others and n not in node.equivalence.nodes:
-                    is_root_cause = False
-                    break
-                if n.result_variable is not None and n.result_variable.hasValue:
-                    continue
-                process_step(self._assignment_step(n))
-                process_step(self._parent_instance_step(n))
-                process_step(self._child_attribute_step(n))
-            if is_root_cause:
-                causes.add(node)
-        return causes
+        roots: Set[AttributeNode] = set(())
+        seen: Set[AttributeNode] = set(())
+        to_check: List[AssignableNode] = self.nodes
+
+        def has_root(node: AssignableNode, cycledetect=set()) -> bool:
+            if node in seen:
+                # Already processed, roots are already in roots
+                return True
+
+            if node.value_assignments or (node.result_variable is not None and node.result_variable.hasValue):
+                # Has a value, never has a root cause
+                return False
+
+            if node in cycledetect:
+                # part of a cycle, cycle in itself doesn't provide a root
+                return False
+
+            sub_cycle = cycledetect.union(node.equivalence.nodes)
+
+            n_has_root = any(
+                (
+                    has_root(subnode, sub_cycle)
+                    for subnode in chain(
+                        self._assignment_step(node), self._parent_instance_step(node), self._child_attribute_step(node)
+                    )
+                )
+            )
+
+            seen.update(node.equivalence.nodes)
+
+            n_is_root = not n_has_root
+
+            if n_is_root:
+                if not node in self.nodes:
+                    # it is root, but not one we are looking for, ignore it
+                    return False
+                else:
+                    roots.update(node.equivalence.nodes)
```

Comment:
```suggestion
                    roots.update(self.nodes.intersection(node.equivalence.nodes))
```
Maybe this should be tested as well.

Reply:
I've added a test to `issue/1821-data-flow-root-cause`. If you merge it in here it should succeed with the suggestion above.